### PR TITLE
Split MuzeroRepresentationLearner out of MuzeroAlgorithm

### PR DESCRIPTION
**Pull request by breakds**
# Motivation

In order to be able to use MuZero's representation learner in other RL algorithms, we will need a standalone representation learner. It basically resembles `MuzeroAlgorithm` but currently it is inconvenient to combine it with another algorithm, say SAC or PPO.

We would like to take a couple of steps to create a standalone MuZero-style representation learner to achieve the above goal.

# What has been done

In this PR, `MuzeroAlgorithm` is split into 2 parts:

1. `MuzeroRepresentationLearner` - It inherits almost all the logic from the original `MuzeroAlgorithm`. It took the ownership of the underlying `MCTSModel`, and is entitled to perform the reanalyze algorithm. `MuzeroRepresentationLearner` trains a model to produce the latent representation from a raw observation.
2. `MuzeroAlgorithm` - It is repurposed to play the role of a wrapper that combines the `MuzeroRepresentationLearner` and the `MCTSAlgorithm`. The user should be able to use `MuzeroAlgorithm` the same as before, except for that they will need to create `MuzeroRepresentationLearner` first.

To accommodate such split, the `initial_inference()` in `MCTSModel` is also divided into two parts.

# Testing

1. Step-by-step comparison of intermediate result on `muzero_atari` for 200 iterations and they are all identical.
2. Identical training curves for `muzero_atari` (on 2 seeds) 
4. Similar performance on `muzero_pendulum` (see figure below).

pendulum_split

Blue = After the split
Orange = Before the split

### CODE REVIEW

**Comment by breakds** [comment]
Found that `git` is not smart enough to identify that `muzero_representation_learner.py` is a clone of `muzero_algorithm.py` before modified. Let me reorganize the commits.

**Comment by breakds** [comment]
Made some arrangement, so there are 2 commits in this PR now. The first commit is a dummy one that just copies `muzero_algorithm.py` into `muzero_representation_learner.py`. The reviewer can read the 2nd commit directly and it will be much easier to see what has been changed.

**Comment by emailweixu** [comment]
> Found that `git` is not smart enough to identify that `muzero_representation_learner.py` is a clone of `muzero_algorithm.py` before modified. Let me reorganize the commits.

Did you try `git mv muzero_algorithm.py muzero_representation_learner.py`? This way git can know the correspondence and move the previous change history to the new file.

**Comment by breakds** [comment]
> > Found that `git` is not smart enough to identify that `muzero_representation_learner.py` is a clone of `muzero_algorithm.py` before modified. Let me reorganize the commits.
> 
> Did you try `git mv muzero_algorithm.py muzero_representation_learner.py`? This way git can know the correspondence and move the previous change history to the new file.



> > Found that `git` is not smart enough to identify that `muzero_representation_learner.py` is a clone of `muzero_algorithm.py` before modified. Let me reorganize the commits.
> 
> Did you try `git mv muzero_algorithm.py muzero_representation_learner.py`? This way git can know the correspondence and move the previous change history to the new file.

I tried `git mv` but could not find a way to make GitHub to be aware of that when combining the two commits. Less optimal way is just to review the 2nd commit: https://github.com/HorizonRobotics/alf/pull/1249/commits/84523f13b70b00f01635cf55468b511c53731728 directly. Sorry!

File: alf/algorithms/muzero_algorithm.py [hunk 1]
```diff
@@ -0,0 +1,163 @@
...
+                 reward_transformer=None,
+                 config: Optional[TrainerConfig] = None,
+                 enable_amp: bool = True,
+                 debug_summaries=False,
+                 name="MuZero"):
+        """
+        Args:
+            observation_spec (TensorSpec): representing the observations.
+            action_spec (BoundedTensorSpec): representing the actions.
+            representation_learner_ctor: will be called to construct a MuZero-style
```
Comment (emailweixu) [inline alf/algorithms/muzero_algorithm.py]: Comment what arguments this ctor expects.

File: alf/algorithms/muzero_algorithm.py [hunk 2]
```diff
@@ -0,0 +1,163 @@
...
+    ``train_step()`` and ``preprocess_experience()`` will delegate directly to their
+    counterparts in the representation learner. This behavior can be changed if needed in
+    the future.
+
+    """
+
+    def __init__(self,
+                 observation_spec,
+                 action_spec,
+                 representation_learner_ctor: Callable[
```
Comment (emailweixu) [inline alf/algorithms/muzero_algorithm.py]: This can be default to MuzeroRepresentationLearner?

File: alf/examples/muzero_tic_tac_toe.gin [hunk 3]
```diff
@@ -17,10 +17,13 @@ MCTSAlgorithm.pb_c_base=19652
 MCTSAlgorithm.is_two_player_game=True
 VisitSoftmaxTemperatureByMoves.move_temperature_pairs=[(0, 1.0), (10, 0.0001)]
 MCTSAlgorithm.visit_softmax_temperature_fn=@VisitSoftmaxTemperatureByMoves()
+
+MuzeroRepresentationLearner.model_ctor=@SimpleMCTSModel
```
Comment (emailweixu) [inline alf/examples/muzero_tic_tac_toe.gin]: muzero_go also needs to be updated.

File: alf/algorithms/muzero_algorithm.py [hunk 4]
```diff
@@ -13,190 +13,100 @@
...
-            observation_spec,
-            action_spec,
-            num_unroll_steps=num_unroll_steps,
-            debug_summaries=debug_summaries)
-        mcts = mcts_algorithm_ctor(
+
+        representation_learner = representation_learner_ctor(
             observation_spec=observation_spec,
             action_spec=action_spec,
+            config=config,
```
Comment (emailweixu) [inline alf/algorithms/muzero_algorithm.py:107]: reward_transformer, reward_spec and enable_amp also need to be passed.

File: alf/algorithms/muzero_representation_learner.py [hunk 5]
```diff
@@ -55,31 +55,48 @@
...
             self,
             observation_spec,
             action_spec,
             model_ctor,
-            mcts_algorithm_ctor,
-            num_unroll_steps,
-            td_steps,
+            num_unroll_steps: int,
+            td_steps: int,
+            discount: float,
```
Comment (emailweixu) [inline alf/algorithms/muzero_representation_learner.py:91]: it's kind of annoying that discount needs to be specified multiple times in MCTSAlgorithm and here.

File: alf/algorithms/mcts_models.py [hunk 6]
```diff
@@ -235,14 +236,38 @@ def __init__(
             self._dynamics_net.set_batch_norm_max_steps(num_unroll_steps)
             self._prediction_net.set_batch_norm_max_steps(num_unroll_steps + 1)
 
-    def initial_inference(self, observation):
-        """Generate the initial prediction given observation.
+    @property
+    def repr_spec(self) -> TensorSpec:
```
Comment (emailweixu) [inline alf/algorithms/mcts_models.py:240]: doc-string

File: alf/algorithms/muzero_algorithm.py (hunk 1 above)
Comment (breakds) [reply alf/algorithms/muzero_algorithm.py]: Fixed.

File: alf/algorithms/muzero_algorithm.py (hunk 2 above)
Comment (breakds) [reply alf/algorithms/muzero_algorithm.py]: Fixed.

File: alf/algorithms/muzero_algorithm.py (hunk 4 above)
Comment (breakds) [reply alf/algorithms/muzero_algorithm.py:107]: Fixed.

File: alf/examples/muzero_tic_tac_toe.gin (hunk 3 above)
Comment (breakds) [reply alf/examples/muzero_tic_tac_toe.gin]: Updated.

File: alf/algorithms/muzero_representation_learner.py (hunk 5 above)
Comment (breakds) [reply alf/algorithms/muzero_representation_learner.py:91]: Right. As a mitigation here I put it in `MuzeroAlgorithm` and asked it to distribute `discount` to both `MCTSAlgorithm` and `MuzeroRepresentationImpl`. This is not the case for `Agent` though.

File: alf/algorithms/mcts_models.py (hunk 6 above)
Comment (breakds) [reply alf/algorithms/mcts_models.py:240]: Added.

**Review by emailweixu — APPROVED** [review]

**Pull request merged**